### ISSUE TITLE
Feature request: allowOutsideClick for keyboard events

### ISSUE THREAD

**Issue by zioth** [focus-trap/focus-trap#199]
`allowOutsideClick` can be used to allow certain page elements to be clickable outside the focus trap. However, these elements do not receive keyboard events, including tabs, which is an accessibility issue. My feature request is either to pass keyboard events and `tabbable` filtering through `allowOutsideClick`, or else create an `allowOutsideKeystrokes` option.

**Comment by stefcameron** [comment]
Thanks for your thoughts on this option. The goal of this library is to trap focus within a certain "boundary" or "scope" element. The mouse (and our fingers, in the case of touch) has an advantage over the keyboard in that it's possible to click/touch an element __without__ letting focus change to that element, hence the name of this option to allow outside _clicks_.

Unfortunately for the keyboard, I don't believe there's a way to direct keystrokes to an element _without_ first setting focus to that element. Adding an `allowOutsideKeystrokes` option sounds like it would violate the very nature of a "focus trap".

The point is to trap focus inside a container element, not letting it get outside of it. If you need keyboard focus to get to some other element outside the trap, perhaps there's a design flaw that needs to be fixed?

**Comment by zioth** [comment]
What you're saying makes perfect sense from a philosophical perspective - if a focus trap doesn't always trap focus, then it's not what it says it is. However, there are cases where you want a specific element to be interactable, even though it's outside the trap.

Here's an example: I open a dialog, and trap focus to that dialog. While the dialog is still open, the client's messaging system (which is not a child node of the dialog) pops up an error. The user cannot dismiss the error, unless you use the function version of `allowOutsideClick` to create an exception, but even with that, keyboard users still can't dismiss it. This creates both an accessibility issue, and some UX confusion, because `:active` and `:focus` states are never set.

Rather than my initial `allowOutsideKeystrokes` suggestion, which is a bit awkward, what about an `isException` callback, which behaves like this?
```
isException: e => e.target.id = 'error_message'
```
It could be called in every click and keyboard event, as well as being used to calculate the list of tabbable elements.

**Comment by stefcameron** [comment]
Thank you for that example: It illustrates the need nicely!

We could explore a solution that uses a callback to check if a node should be an exception, as you proposed.

Here's what I'm thinking:

- The [updateTabbableNodes()](https://github.com/focus-trap/focus-trap/blob/da02d77db7a8f8b520eae2ba526fa8b0702b46a6/index.js#L326) function is called in 3 places, including on each TAB key press, to determine the current list of tabbable nodes according to the element that defines the trap (the `container` element provided to the `createFocusTrap()` function).
- The [checkFocusIn()](https://github.com/focus-trap/focus-trap/blob/da02d77db7a8f8b520eae2ba526fa8b0702b46a6/index.js#L274) function checks `container.contains(e.target)` -- this seems like it should rather check that `e.target` is in the current list of known tabbable nodes. Right now, that would be the same as checking the container contains the target, but if we added a callback to modify the list of tabbable nodes, then it could potentially "contain" more (or less) than what `tabbable(container)` determines.
- We add a `tabbableHook(nodes: Array<Element>, container: Element) => Array<Element>` option to the focus trap configuration: This hook gets called by `updateTabbableNodes()`, giving it the list of nodes that `tabbable(container)` found, as well as the container used for the search. The function could then either return the same list (default) or return an augmented (or even diminished) list, including the additional nodes for the notification popup that just appeared (as in your example).

I _think_ that would work, __however__, I'm not sure what to do about the [checkTab()](https://github.com/focus-trap/focus-trap/blob/da02d77db7a8f8b520eae2ba526fa8b0702b46a6/index.js#L297) method, as well as [firstTabbableNode/lastTabbableNode](https://github.com/focus-trap/focus-trap/blob/da02d77db7a8f8b520eae2ba526fa8b0702b46a6/index.js#L328-L329) (set in `updateTabbableNodes()`) because if the list of tabbable nodes now includes nodes that are _outside_ the container, what is the first and what is the last because questionable.

One solution is to leave it up to the hook by simply stating that the first and last tabbable nodes will be determined by the first and last nodes in the array the hook returns (basically as it's done today with the `tabbable(container)` results), then the onus is on the hook to make sure the first/last nodes make logical sense.

There's also the question of what kind of tabbing order the browser will infer once we let focus go to a node that isn't in the container.

Thoughts?

**Comment by zioth** [comment]
That sounds like a good solution. My only concern is that checkFocusIn will become far less performant, since it will have to do a large number of DOM operations (one on each tabbable node) instead of just calling `contains`. This might make an exception list a better option, because then you can call `contains` once for the focus trap, and then once for each exception.

For `tabbableHook`, you may want to use selectors instead of elements (or allow both). The reason is that elements will make it much more difficult to use this in dynamically-generated content, particularly in React, when you don't even have direct access to elements without some extra work.

`checkTab`: One solution would be to call `tabbable` on the whole document instead of the container, then filter based on the container and exceptions in the `tabbable` code. That would guarantee that nodes stay in the correct order. You'd then have to keep multiple lists of `firstNode`/`lastNode` pairs, to make sure you jump from one to the next, rather than allow tabbing between lists. Or you could simplify the operation by doing all tab handling with `tryFocus()`.

I wouldn't depend on the hook to sort tabbable nodes, if it can be avoided. It would require the caller to know too much about the structure of the DOM.

**Comment by stefcameron** [comment]
I believe #217 will fix this issue, so I've marked it as such.

**Comment by stefcameron** [comment]
@all-contributors please add @zioth for ideas

Thanks for suggesting this!
